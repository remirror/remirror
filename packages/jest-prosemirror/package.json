{
  "name": "jest-prosemirror",
  "version": "3.0.0-beta.4",
  "description": "Write expressive tests for your prosemirror editor",
  "keywords": [
    "jest",
    "prosemirror",
    "snapshots",
    "tests"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-prosemirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-prosemirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-prosemirror.js",
      "require": "./dist/jest-prosemirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-prosemirror-environment.js",
      "require": "./environment/dist/jest-prosemirror-environment.cjs"
    },
    "./package.json": "./package.json",
    "./serializer": {
      "types": "./dist-types/serializer.d.ts",
      "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
    }
  },
  "main": "./dist/jest-prosemirror.cjs",
  "module": "./dist/jest-prosemirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment",
    "serializer"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core-constants": "3.0.0-beta.3",
    "@remirror/core-helpers": "4.0.0-beta.3",
    "@remirror/core-types": "3.0.0-beta.4",
    "@remirror/core-utils": "3.0.0-beta.4",
    "@remirror/pm": "3.0.0-beta.4",
    "@testing-library/dom": "^9.3.0",
    "@types/stringify-object": "^3.3.1",
    "chalk": "^4.1.0",
    "jest-matcher-utils": "^29.5.0",
    "prosemirror-schema-basic": "^1.2.3",
    "prosemirror-tables": "^1.3.7",
    "prosemirror-test-builder": "^1.1.1",
    "stringify-object": "^3.3.0",
    "test-keyboard": "2.0.7-beta.4"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-prosemirror.d.ts",
        "import": "./dist/jest-prosemirror.js",
        "require": "./dist/jest-prosemirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-prosemirror-environment.d.ts",
        "import": "./environment/dist/jest-prosemirror-environment.js",
        "require": "./environment/dist/jest-prosemirror-environment.cjs"
      },
      "./package.json": "./package.json",
      "./serializer": {
        "types": "./serializer/dist/jest-prosemirror-serializer.d.ts",
        "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
      }
    },
    "main": "./dist/jest-prosemirror.cjs",
    "module": "./dist/jest-prosemirror.js",
    "type": "module",
    "types": "./dist/jest-prosemirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts",
      "serializer.cts"
    ]
  }
}
