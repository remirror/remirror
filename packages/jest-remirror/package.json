{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
