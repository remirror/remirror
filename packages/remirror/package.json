{
  "name": "remirror",
  "version": "3.0.0-beta.8",
  "description": "One editing package to rule them all, one editing package to bind them.",
  "keywords": [
    "prosemirror",
    "react",
    "remirror",
    "wysiwyg",
    "editing",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>",
    "ocavue "
  ],
  "sideEffects": false,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/remirror.js",
      "require": "./dist/remirror.cjs"
    },
    "./dom": {
      "types": "./dist-types/dom.d.ts",
      "import": "./dom/dist/remirror-dom.js",
      "require": "./dom/dist/remirror-dom.cjs"
    },
    "./extensions": {
      "types": "./dist-types/extensions.d.ts",
      "import": "./extensions/dist/remirror-extensions.js",
      "require": "./extensions/dist/remirror-extensions.cjs"
    },
    "./icons": {
      "types": "./dist-types/icons.d.ts",
      "import": "./icons/dist/remirror-icons.js",
      "require": "./icons/dist/remirror-icons.cjs"
    },
    "./package.json": "./package.json",
    "./styles/*": "./styles/*"
  },
  "main": "./dist/remirror.cjs",
  "module": "./dist/remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "dom",
    "extensions",
    "icons",
    "styles"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.8",
    "@remirror/core-constants": "3.0.0-beta.4",
    "@remirror/core-helpers": "4.0.0-beta.5",
    "@remirror/core-types": "3.0.0-beta.5",
    "@remirror/core-utils": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.8",
    "@remirror/extension-annotation": "3.0.0-beta.8",
    "@remirror/extension-bidi": "3.0.0-beta.8",
    "@remirror/extension-blockquote": "3.0.0-beta.8",
    "@remirror/extension-bold": "3.0.0-beta.8",
    "@remirror/extension-callout": "3.0.0-beta.8",
    "@remirror/extension-code": "3.0.0-beta.8",
    "@remirror/extension-code-block": "3.0.0-beta.8",
    "@remirror/extension-collaboration": "3.0.0-beta.8",
    "@remirror/extension-columns": "3.0.0-beta.8",
    "@remirror/extension-diff": "3.0.0-beta.8",
    "@remirror/extension-doc": "3.0.0-beta.8",
    "@remirror/extension-drop-cursor": "3.0.0-beta.8",
    "@remirror/extension-embed": "3.0.0-beta.8",
    "@remirror/extension-emoji": "3.0.0-beta.8",
    "@remirror/extension-entity-reference": "3.0.0-beta.8",
    "@remirror/extension-epic-mode": "3.0.0-beta.8",
    "@remirror/extension-events": "3.0.0-beta.8",
    "@remirror/extension-find": "1.0.0-beta.8",
    "@remirror/extension-font-family": "3.0.0-beta.8",
    "@remirror/extension-font-size": "3.0.0-beta.8",
    "@remirror/extension-gap-cursor": "3.0.0-beta.8",
    "@remirror/extension-hard-break": "3.0.0-beta.8",
    "@remirror/extension-heading": "3.0.0-beta.8",
    "@remirror/extension-history": "3.0.0-beta.8",
    "@remirror/extension-horizontal-rule": "3.0.0-beta.8",
    "@remirror/extension-image": "3.0.0-beta.8",
    "@remirror/extension-italic": "3.0.0-beta.8",
    "@remirror/extension-link": "3.0.0-beta.8",
    "@remirror/extension-list": "3.0.0-beta.8",
    "@remirror/extension-markdown": "3.0.0-beta.8",
    "@remirror/extension-mention": "3.0.0-beta.8",
    "@remirror/extension-mention-atom": "3.0.0-beta.8",
    "@remirror/extension-node-formatting": "3.0.0-beta.8",
    "@remirror/extension-paragraph": "3.0.0-beta.8",
    "@remirror/extension-placeholder": "3.0.0-beta.8",
    "@remirror/extension-positioner": "3.0.0-beta.8",
    "@remirror/extension-shortcuts": "3.0.0-beta.8",
    "@remirror/extension-strike": "3.0.0-beta.8",
    "@remirror/extension-sub": "3.0.0-beta.8",
    "@remirror/extension-sup": "3.0.0-beta.8",
    "@remirror/extension-tables": "3.0.0-beta.8",
    "@remirror/extension-text": "3.0.0-beta.8",
    "@remirror/extension-text-case": "3.0.0-beta.8",
    "@remirror/extension-text-color": "3.0.0-beta.8",
    "@remirror/extension-text-highlight": "3.0.0-beta.8",
    "@remirror/extension-trailing-node": "3.0.0-beta.8",
    "@remirror/extension-underline": "3.0.0-beta.8",
    "@remirror/extension-whitespace": "3.0.0-beta.8",
    "@remirror/icons": "3.0.0-beta.5",
    "@remirror/preset-core": "3.0.0-beta.8",
    "@remirror/preset-formatting": "3.0.0-beta.8",
    "@remirror/preset-wysiwyg": "3.0.0-beta.8",
    "@remirror/theme": "3.0.0-beta.5",
    "@types/refractor": "^3.0.2",
    "refractor": "^3.6.0"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.1",
    "@remirror/pm": "3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependencies": {
    "@remirror/pm": "^3.0.0-beta.6",
    "@types/prettier": "^2.7.2",
    "prettier": "^2.8.8"
  },
  "peerDependenciesMeta": {
    "@types/prettier": {
      "optional": true
    },
    "prettier": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/remirror.d.ts",
        "import": "./dist/remirror.js",
        "require": "./dist/remirror.cjs"
      },
      "./dom": {
        "types": "./dom/dist/remirror-dom.d.ts",
        "import": "./dom/dist/remirror-dom.js",
        "require": "./dom/dist/remirror-dom.cjs"
      },
      "./extensions": {
        "types": "./extensions/dist/remirror-extensions.d.ts",
        "import": "./extensions/dist/remirror-extensions.js",
        "require": "./extensions/dist/remirror-extensions.cjs"
      },
      "./icons": {
        "types": "./icons/dist/remirror-icons.d.ts",
        "import": "./icons/dist/remirror-icons.js",
        "require": "./icons/dist/remirror-icons.cjs"
      },
      "./package.json": "./package.json",
      "./styles/*": "./styles/*"
    },
    "main": "./dist/remirror.cjs",
    "module": "./dist/remirror.js",
    "type": "module",
    "types": "./dist/remirror.d.ts"
  },
  "@remirror": {},
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "dom.ts",
      "extensions.ts",
      "icons.ts"
    ]
  }
}
