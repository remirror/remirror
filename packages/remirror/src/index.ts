export * from '@remirror/core';
export * from '@remirror/theme';
