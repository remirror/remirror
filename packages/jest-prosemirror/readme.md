# jest-prosemirror

[![npm](https://img.shields.io/npm/dm/jest-prosemirror.svg?&logo=npm)](https://www.npmjs.com/package/jest-prosemirror)

## The problem

You want to write tests for some of your prosemirror editor but you don't know where to start. You know you should avoid testing implementation details and just want to be sure that your commands and plugins produce the correct underlying prosemirror state.

## This solution

`jest-prosemirror` takes inspiration from the [`testing-library`](https://github.com/testing-library/react-testing-library) mantra and enables you to write more intuitive tests for your prosemirror editor.

## Installation

```bash
yarn add jest-prosemirror # yarn
pnpm add jest-prosemirror # pnpm
npm install jest-prosemirror # npm
```

## Getting started

### Quick setup

For a quick setup add the following to your jest.config.js file.

```js
module.exports = {
  setupFilesAfterEnv: ['jest-prosemirror/environment'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

This will automatically:

- Add the jest assertions `toTransformNode` and `toEqualProsemirrorNode`.

If you are using typescript then add this to your `tsconfig.json` file for global type support.

```json
{
  "compilerOptions": {
    "types": ["jest-prosemirror"]
  }
}
```

### Manual setup

Create a `jest.framework.dom.ts` file and add the following

```ts
// jest.framework.dom.ts

import { prosemirrorMatchers } from 'jest-prosemirror';

// Add jest-prosemirror assertions
expect.extend(prosemirrorMatchers);
```

In your `jest.config.js` file add this to the configuration

```js
module.exports = {
  setupFilesAfterEnv: ['<rootDir>/jest.framework.dom.ts'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

## Snapshot serializer

This package exports a serializer for better snapshot testing of prosemirror primitives. To set this up add the following to your `jest.config.js` file.

```js
module.exports = {
  snapshotSerializers: ['jest-prosemirror/serializer'],
};
```

Alternatively, you can add the following to your `jest.framework.dom.ts` file.

```ts
// jest.framework.dom.ts

import { prosemirrorSerializer } from 'jest-prosemirror';

// Add the serializer for use throughout all the configured test files.
expect.addSnapshotSerializer(prosemirrorSerializer);
```

## Jest Matchers

### `toTransformNode`

A utility from jest-prosemirror which tests that a command transforms the prosemirror node in the desired way.

```ts
import { doc, p, schema, strong } from 'jest-prosemirror';
import { removeMark } from 'remirror/core/utils';

test('remove the mark', () => {
  const type = schema.marks.bold;
  const from = doc(p(strong('<start>bold<end>')));
  const to = doc(p('bold'));

  expect(removeMark({ type })).toTransformNode({ from, to });
});
```

This tests that mark has been removed by the provided command.

The `to` property is optional and can be left blank to test that the node is identical after the transform.

### `toEqualProsemirrorNode`

Tests that two prosemirror documents are equal. Pass in the expected document and it checks that they are the same.

```ts
import { createEditor, doc, p } from 'jest-prosemirror';
import { removeNodeAtPosition } from 'remirror/core/utils';

test('remove block top level node at specified position', () => {
  const {
    state: { tr },
  } = createEditor(doc(p('x'), p('one')));
  const newTr = removeNodeAtPosition({ pos: 3, tr });

  expect(newTr).not.toBe(tr);
  expect(newTr.doc).toEqualProsemirrorNode(doc(p('x')));
});
```

## `createEditor`

Create a test prosemirror editor.

The call to create editor can be chained with various commands to enable testing of the editor at each step without the need for intermediate holding variables.

```ts
import { createEditor, doc, p } from 'jest-remirror';
import { suggest } from 'prosemirror-suggest';

test('`keyBindings`', () => {
  const keyBindings = {
    Enter: jest.fn((params: SuggestKeyBindingParameter) => {
      params.command();
    }),
  };

  const plugin = suggest({
    char: '@',
    name: 'at',
    keyBindings,
    matchOffset: 0,
    createCommand:
      ({ view }) =>
      () =>
        view.dispatch(view.state.tr.insertText('awesome')),
  });

  createEditor(doc(p('<cursor>')), { plugins: [plugin] })
    .insertText('@')
    .press('Enter')
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorNode(doc(p('@awesome')));
    });
});
```

### Parameters

#### taggedDoc - `TaggedProsemirrorNode`

The tagged prosemirror node to inject into the editor.

#### options - `CreateEditorOptions`

The following options which include all [DirectEditorProps](http://prosemirror.net/docs/ref/#view.DirectEditorProps) except for `state`.

| **Property** | **Type** | **Default** | **Description** |
| --- | :-: | :-: | --- |
| `autoClean` | `boolean` | `true` | Whether to auto remove the editor from the dom after each test. It is advisable to leave this unchanged. |
| `plugins` | `Plugin[]` | `[]` | The plugins that the test editor should use. |
| `rules` | `InputRule[]` | `[]` | The input rules that the test editor should use. |

### Returns

#### start - `number`

The start of the current selection.

#### end - `number | undefined`

The end of the current selection.

#### selection - `Selection`

The current prosemirror selection

#### schema - `EditorSchema`

The prosemirror schema.

#### state - `EditorState`

The prosemirror state.

#### view - `EditorView`

The prosemirror view.

#### overwrite - `(doc: TaggedProsemirrorNode) => ReturnType<typeof createEditor>`

Overwrite all the current content within the editor.

**`doc`** - the new content to use.

#### command - `(command: CommandFunction) => ReturnType<typeof createEditor>`

Run the command within the prosemirror editor.

```ts
test('commands are run', () => {
  createEditor(doc(p('<cursor>')))
    .command((state, dispatch) => {
      if (dispatch) {
        dispatch(state.tr.insertText('hello'));
      }
    })
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorDocument(doc(p('hello')));
    });
});
```

**`command`** - the command function to run.

#### insertText - `(text: string) => ReturnType<typeof createEditor>`

Insert text into the editor at the current position.

**`text`** - the text to insert

#### jumpTo - `(start: 'start' | 'end' | number, end?: number) => ReturnType<typeof createEditor>`

Jump to the specified position in the editor.

**`start`** - a number position or the shorthand 'start' | 'end'. **`[end]`** - the option end position of the new selection.

#### shortcut - `(mod: string) => ReturnType<typeof createEditor>`

Type a keyboard shortcut - e.g. `Mod-Enter`.

**NOTE** This only simulates the events. For example an `Mod-Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`mod`** - the keyboard shortcut to type

#### press - `(char: string) => ReturnType<typeof createEditor>`

Simulate a keypress which is run through the editor's key handlers.

**NOTE** This only simulates the events. For example an `Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`char`** - the character to type

#### fire - `(params: Omit<FireEventAtPositionParameter, 'view'>) => ReturnType<typeof createEditor>`

Fire an event in the editor. This api is not the most well tested and can be quite flakey.

**`params`** - the fire event parameters

#### callback - `(fn: (content: ReturnValueCallbackParameter) => void) => ReturnType<typeof createEditor>`

Callback function which receives the `start`, `end`, `state`, `view`, `schema` and `selection` properties and allows for easier testing of the current state of the editor.

## Acknowledgements

This package borrows very heavily from [prosemirror-test-builder](https://www.npmjs.com/package/prosemirror-test-builder)
