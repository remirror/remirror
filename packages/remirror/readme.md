# remirror

> One package to rule them all, one entry point to bind them.

[![Version][version]][npm] [![Weekly Downloads][downloads-badge]][npm] [![Bundled size][size-badge]][size] [![Typed Codebase][typescript]](#) [![MIT License][license]](#)

[version]: https://flat.badgen.net/npm/v/remirror
[npm]: https://npmjs.com/package/remirror
[license]: https://flat.badgen.net/badge/license/MIT/purple
[size]: https://bundlephobia.com/result?p=remirror
[size-badge]: https://flat.badgen.net/bundlephobia/minzip/remirror
[typescript]: https://flat.badgen.net/badge/icon/TypeScript?icon=typescript&label
[downloads-badge]: https://badgen.net/npm/dw/remirror/red?icon=npm

## Installation

```bash
# yarn
yarn add remirror @remirror/pm

# pnpm
pnpm add remirror @remirror/pm

# npm
npm install remirror @remirror/pm
```

The `remirror` package will automatically install the `@remirror/pm` package for you. You only need to install it yourself if you'd like to ensure consistent versions of the `prosemirror-*` libraries when importing from `@remirror/pm/state` instead of `prosemirror-state` or `@remirror/pm/model` instead of `prosemirror-model`.

## Usage

Rather than installing multiple scoped packages, the `remirror` package is a gateway to using all the goodness that remirror provides while minimising your bundle size.

The following creates a controlled editor with React.

```tsx
import React from 'react';
import { SocialPreset } from 'remirror/extensions';
import { Remirror, SocialEmojiComponent, useRemirror } from '@remirror/react';

const EditorWrapper = () => {
  const socialPreset = new SocialPreset();
  const { state, onChange } = useRemirror({ extensions: () => [...socialPreset()] });

  return (
    <Remirror state={state} onChange={onChange} manager={manager} autoRender={true}>
      <SocialEmojiComponent />
    </Remirror>
  );
};
```

These are the entry points available through the `remirror` package.

- `remirror` - All the core functionality available through `@remirror/core`.
- `remirror/extensions` - All the core extensions and presets made available through the main `remirror` repository. This doesn't include any framework specific extensions and presets.
- `remirror/dom` - The dom framework implementation of via `createDomEditor`.
