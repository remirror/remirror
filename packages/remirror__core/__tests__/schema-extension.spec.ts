import { jest } from '@jest/globals';
import { schema } from 'jest-prosemirror';
import { extensionValidityTest, renderEditor } from 'jest-remirror';
import { values } from 'remirror';
import {
  BlockquoteExtension,
  BoldExtension,
  corePreset,
  HeadingExtension,
  wysiwygPreset,
} from 'remirror/extensions';

import {
  ApplySchemaAttributes,
  ExtensionTag,
  MarkExtension,
  NodeExtension,
  PlainExtension,
  RemirrorManager,
  SchemaExtension,
} from '../';

extensionValidityTest(SchemaExtension);

describe('dynamic schema attributes', () => {
  it('should dynamically add attributes when configured', () => {
    let paragraphId = 0;
    let markId = 0;

    const mark = {
      paragraph: jest.fn(() => `${++paragraphId}`),
      marks: jest.fn(() => `${++markId}`),
    };

    const editor = renderEditor(
      [new HeadingExtension(), new BlockquoteExtension(), new BoldExtension()],
      {
        extraAttributes: [
          { identifiers: ['paragraph'], attributes: { id: { default: mark.paragraph } } },
          { identifiers: 'marks', attributes: { id: mark.marks } },
        ],
      },
    );

    const { doc, heading: h } = editor.nodes;

    editor.add(doc(h('This is a heading<cursor>')));
    expect(mark.paragraph).not.toHaveBeenCalled();
    expect(mark.marks).not.toHaveBeenCalled();

    editor.insertText('\nWelcome to the **bold** experiment\n\n\n');
    expect(mark.marks).toHaveBeenCalledTimes(1);
    expect(mark.paragraph).toHaveBeenCalledTimes(4);

    const outerHtml = editor.dom.outerHTML;
    expect(editor.dom.outerHTML).toMatchSnapshot();

    editor.selectText(1);

    // Nothing should have changed.
    expect(mark.marks).toHaveBeenCalledTimes(1);
    expect(mark.paragraph).toHaveBeenCalledTimes(4);
    expect(outerHtml).toMatchSnapshot(editor.dom.outerHTML);
  });
});

test('custom schema', () => {
  const editor = renderEditor<never>([], { schema });

  expect(editor.schema).toBe(schema);
  expect(Object.keys(editor.manager.nodes)).toMatchSnapshot();
  expect(Object.keys(editor.manager.marks)).toMatchSnapshot();
});

describe('extraAttributes', () => {
  it('should support adding attributes to `nodes`, `marks`, and `all`', () => {
    const manager = RemirrorManager.create(() => [...wysiwygPreset(), ...corePreset()], {
      extraAttributes: [
        { identifiers: 'nodes', attributes: { totallyNodes: 'abc' } },
        { identifiers: 'marks', attributes: { totallyMarks: 'abc' } },
        { identifiers: 'all', attributes: { totallyAll: 'abc' } },
      ],
    });

    for (const type of values(manager.schema.nodes)) {
      if (['doc', 'text'].includes(type.name)) {
        continue;
      }

      expect(type.spec.attrs?.totallyNodes).toEqual({ default: 'abc' });
      expect(type.spec.attrs).not.toHaveProperty('totallyMarks');
      expect(type.spec.attrs?.totallyAll).toEqual({ default: 'abc' });
    }

    for (const type of values(manager.schema.marks)) {
      expect(type.spec.attrs?.totallyMarks).toEqual({ default: 'abc' });
      expect(type.spec.attrs).not.toHaveProperty('totallyNodes');
      expect(type.spec.attrs?.totallyAll).toEqual({ default: 'abc' });
    }
  });

  it('should support adding attributes by name', () => {
    const manager = RemirrorManager.create(() => [...wysiwygPreset(), ...corePreset()], {
      extraAttributes: [
        { identifiers: ['bold'], attributes: { totallyBold: 'abc' } },
        { identifiers: ['paragraph', 'italic'], attributes: { fun: 'abc' } },
      ],
    });

    expect(manager.schema.nodes.paragraph.spec.attrs?.fun).toEqual({ default: 'abc' });
    expect(manager.schema.marks.italic.spec.attrs?.fun).toEqual({ default: 'abc' });
    expect(manager.schema.marks.bold.spec.attrs).not.toHaveProperty('fun');
    expect(manager.schema.marks.bold.spec.attrs?.totallyBold).toEqual({ default: 'abc' });
  });

  it('should support adding attributes by tag', () => {
    // Here to ensure no errors are caused when a plain extension is found.
    class ThePlainExtension extends PlainExtension {
      get name() {
        return 'thePlain' as const;
      }

      createTags() {
        return [ExtensionTag.Alignment];
      }
    }

    class TheMarkExtension extends MarkExtension {
      get name() {
        return 'theMark' as const;
      }

      createTags() {
        return [ExtensionTag.Alignment, ExtensionTag.Color];
      }

      createMarkSpec(extra: ApplySchemaAttributes) {
        return {
          attrs: extra.defaults(),
        };
      }
    }

    class TheNodeExtension extends NodeExtension {
      get name() {
        return 'theNode' as const;
      }

      createTags() {
        return [ExtensionTag.Alignment];
      }

      createNodeSpec(extra: ApplySchemaAttributes) {
        return {
          attrs: extra.defaults(),
        };
      }
    }

    const manager = RemirrorManager.create(
      () => [
        ...wysiwygPreset(),
        ...corePreset(),
        new ThePlainExtension(),
        new TheMarkExtension(),
        new TheNodeExtension(),
      ],
      {
        extraAttributes: [
          // Matches everything
          {
            identifiers: { tags: [ExtensionTag.Alignment] },
            attributes: { matchesEverything: 'abc' },
          },

          // Only marks
          {
            identifiers: { tags: [ExtensionTag.Alignment], type: 'mark' },
            attributes: { onlyMarks: 'abc' },
          },

          // No matches
          { identifiers: { tags: [] }, attributes: { emptyTags: 'abc' } },

          // No matches - too exclusive
          {
            identifiers: { tags: [ExtensionTag.Color], type: 'node' },
            attributes: { tooExclusive: 'abc' },
          },
        ],
      },
    );

    expect(manager.schema.nodes.theNode.spec.attrs?.matchesEverything).toEqual({
      default: 'abc',
    });
    expect(manager.schema.nodes.theNode.spec.attrs).not.toHaveProperty('onlyMarks');
    expect(manager.schema.nodes.theNode.spec.attrs).not.toHaveProperty('emptyTags');
    expect(manager.schema.nodes.theNode.spec.attrs).not.toHaveProperty('tooExclusive');

    expect(manager.schema.marks.theMark.spec.attrs?.matchesEverything).toEqual({
      default: 'abc',
    });
    expect(manager.schema.marks.theMark.spec.attrs?.onlyMarks).toEqual({ default: 'abc' });
    expect(manager.schema.marks.theMark.spec.attrs).not.toHaveProperty('emptyTags');
    expect(manager.schema.marks.theMark.spec.attrs).not.toHaveProperty('tooExclusive');
  });
});
