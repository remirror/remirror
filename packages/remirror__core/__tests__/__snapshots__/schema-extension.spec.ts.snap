// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom schema 1`] = `
[
  "doc",
  "orderedList",
  "bulletList",
  "listItem",
  "paragraph",
  "text",
  "horizontalRule",
  "atomInline",
  "atomBlock",
  "atomContainer",
  "containerWithRestrictedContent",
  "table",
  "table_row",
  "table_cell",
  "table_header",
  "blockquote",
  "heading",
  "code_block",
  "hard_break",
  "image",
]
`;

exports[`custom schema 2`] = `
[
  "link",
  "em",
  "strong",
  "code",
  "strike",
]
`;

exports[`dynamic schema attributes should dynamically add attributes when configured 1`] = `
<div
  contenteditable="true"
  translate="no"
  role="textbox"
  aria-multiline="true"
  aria-label
  class="ProseMirror remirror-editor"
>
  <h1>
    This is a heading
    <br class="ProseMirror-trailingBreak">
  </h1>
  <p id="1">
    Welcome to the
    <strong id="1">
      bold experiment
    </strong>
    <br class="ProseMirror-trailingBreak">
  </p>
  <p id="2">
    <br class="ProseMirror-trailingBreak">
  </p>
  <p id="3">
    <br class="ProseMirror-trailingBreak">
  </p>
  <p id="4">
    <br class="ProseMirror-trailingBreak">
  </p>
</div>
`;

exports[`dynamic schema attributes should dynamically add attributes when configured: <div contenteditable="true" translate="no" role="textbox" aria-multiline="true" aria-label="" class="ProseMirror remirror-editor"><h1>This is a heading
<br class="ProseMirror-trailingBreak"></h1><p id="1">Welcome to the <strong id="1">bold experiment
</strong><br class="ProseMirror-trailingBreak"></p><p id="2">
<br class="ProseMirror-trailingBreak"></p><p id="3">
<br class="ProseMirror-trailingBreak"></p><p id="4"><br class="ProseMirror-trailingBreak"></p></div> 1`] = `
<div
  contenteditable="true"
  translate="no"
  role="textbox"
  aria-multiline="true"
  aria-label
  class="ProseMirror remirror-editor"
>
  <h1>
    This is a heading
    <br class="ProseMirror-trailingBreak">
  </h1>
  <p id="1">
    Welcome to the
    <strong id="1">
      bold experiment
    </strong>
    <br class="ProseMirror-trailingBreak">
  </p>
  <p id="2">
    <br class="ProseMirror-trailingBreak">
  </p>
  <p id="3">
    <br class="ProseMirror-trailingBreak">
  </p>
  <p id="4">
    <br class="ProseMirror-trailingBreak">
  </p>
</div>
`;
